modules/apps/portal/portal-cluster-multiple/src/main/java/com/liferay/portal/cluster/multiple/internal/jgroups/JGroupsClusterChannelFactory.java

```diff
@@ -247,12 +253,37 @@ protected void setProps(Props props) {
 		_props = props;
 	}
 
+	private InputStream _getConfigStream(String properties) throws IOException {
```

Comment:
@tinatian this loading locations fallback is too flexible. We should not go that far. We should restrict it a bit to avoid unnecessary upgrade problems in the future.
And the parameter name needs to be updated, here we call it "properties", but at the top level, ClusterExecutorImpl.initialize(), we still call it "channelPropertiesString".

Reply:
Hi @dantewang,
Please see Shuyang's comment, fix the issues mentioned, squash the commits when possible, and resend a pull to me. Since we need to backport this to several versions, I would like the commits to be simple and clear.

For the path finding logic, let's do this:
Check current classloader first, if not found, check portal classloader

This is what we did for ehcache config file right now, and it is also doable for 6.2. With this change, you can modify the config file by pointing to a file in "webapps/ROOT/WEB-INF/classes" on every version of portal, also you can use fragment to modify the config on newer versions.

Thanks.
Tina.

Reply:
@tinatian 

> For the path finding logic, let's do this:
> Check current classloader first, if not found, check portal classloader

To confirm, does that mean we drop support treating the path as filesystem path or URL? If so, I have a concern because these are what JGroups itself supports even in the current version (3.16), which means there may be customer using them.

Reply:
Emm.... yes, right, if we drop those, customers may end up modifying their portal-ext.properties after upgrade.
@shuyangzhou  What do you think?

Tina.

Reply:
I thought we are only feeding jgroups raw property string. You mean the current property string can actually be a local file path or even an URL?

Reply:
@shuyangzhou 
Yes, here are the current comments for this property:
`    #
    # Set the JGroups properties for each channel, we support up to 10 transport
    # channels and 1 single required control channel. Use as few transport
    # channels as possible for best performance. By default, only one UDP
    # control channel and one UDP transport channel are enabled. Channels can be
    # configured by XML files that are located in the class path or by inline
    # properties.`
Even though we only say you can config XML files, but in fact, I think it will also work for URL, cause we are depending on the JGroups support for this.

Reply:
@shuyangzhou We've been always feeding the property string into JGroups' channel constructor, inside which JGroups determines whether the string is a raw property string, a filesystem path to an XML file, a URL for an XML file, or a class path resource path to an XML file.

JGroups 4 just removed raw string but kept all other 3.

Although we just documented about the class path part, JGroups documented them all, and users were actually able to set

```
cluster.link.channel.properties.control=${liferay.home}/somepath/UDP.xml
```

And I remember I did see this way when checking some customer issues with CSE.

Reply:
@shuyangzhou 
I just confirmed with Emma that many customers write filesystem path for "cluster.link.channel.properties.*".

By the way, previously I inlined the path lookup logic from JGroups because I intended to put look up in Portal class loader before JGroups class loader. If we put Portal class loader look up last, then all others can be delegated back to JGroups' `ConfigurationFactory` itself. This will make our code look cleaner.